include/dali/core/util.h

```diff
@@ -125,26 +125,83 @@ static_assert(align_up(17, 16) == 32, "Should align up");
 static_assert(align_up(8, 8) == 8, "Should be already aligned");
 static_assert(align_up(5, 8) == 8, "Should align");
 
+/**
+ * @brief Calculates the smallest power of 2 that is greater than or equal to n.
+ *
+ * For n <= 1 the result is 1.  
+ * If n is already a power of 2, the result is n itself.
+ *
+ * @return The least power of 2 ≥ n.
+ */
 template <typename T>
 DALI_HOST_DEV DALI_FORCEINLINE
 constexpr std::enable_if_t<std::is_integral<T>::value, T> next_pow2(T n) {
-  T pow2 = 1;
-  while (n > pow2) {
-    pow2 += pow2;
+  using U = std::make_unsigned_t<T>;
+  U x = static_cast<U>(n);
+
+  if (x <= 1)
+    return 1;
+
+#if defined(__CUDA_ARCH__)
+  // CUDA DEVICE PATH
+  if constexpr (sizeof(U) == 4) {
```

Comment:
Please swap the condition. Now the 64-bit variant would be used for 8 and 16-bit inputs, which is a big waste of resources. Use 64-bit `__clzll` only for 64-bit inputs and 32-bit otherwise.